### PR TITLE
Template the GridFuncVector class with memory space

### PR DESCRIPTION
**Pull request by wjge**
It is better to review and merge these changes now before the PR becomes even larger.

### CODE REVIEW

File: src/pb/FDoper.h [hunk 1]
```diff
@@ -120,6 +120,7 @@ class FDoper : public FDoperInterface
             delete lower_order_grid_;
             lower_order_grid_ = nullptr;
         }
+        del2_4th_tm_.print(std::cout);
```
Comment (jeanlucf22) [inline src/pb/FDoper.h]: This call should not be here. WE print timers together at the end of run.

File: src/pb/FDoper.cc [hunk 2]
```diff
@@ -507,7 +505,7 @@ void FDoper<T>::del2_4th(const Grid& Agrid, T* A, T* B, const size_t nfunc,
     int incx = incx_;
     int incy = incy_;
 
-    MGMOL_PARALLEL_FOR_COLLAPSE(4, A, B)
+    MGMOL_PARALLEL_FOR_COLLAPSE(2, A, B) // 0.037 s 3000x32^3 0.0156 1000x32^3
```
Comment (jeanlucf22) [inline src/pb/FDoper.cc]: should it be in a different PR? It has nothing to do with that one...

File: src/pb/GridFuncVector.cc [hunk 3]
```diff
@@ -495,7 +663,28 @@ void GridFuncVector<ScalarType>::finishNorthSouthComm()
...
+#ifdef HAVE_OPENMP_OFFLOAD
+            MGMOL_PARALLEL_FOR_COLLAPSE(2, functions_alias)
+            for (int color = 0; color < nfunc; color++)
+            {
+                for (int j = 0; j < nghosts * incy; j += incy)
+                {
+                    for (int i = nghosts * incx; i < (dimx + nghosts) * incx;
+                         i += incx)
+                    {
+                        for (int k = 0; k < sdimz; k++)
```
Comment (jeanlucf22) [inline src/pb/GridFuncVector.cc]: this should be dimz, not sdimz

File: src/pb/FDoper.h (hunk 1 above)
Comment (wjge) [reply src/pb/FDoper.h]: I will delete this line in the next commit.

File: src/pb/GridFuncVector.cc (hunk 3 above)
Comment (wjge) [reply src/pb/GridFuncVector.cc]: This bug is fixed.

**Comment by jeanlucf22** [comment]
@wjge The test "testProjectedMatrices" involves functionalities we do not want to support on GPU at the moment. So let's turn it off. It fails because the trade ghosts function is incorrect for that test without the "gids".

File: src/tools/memory_space.h [hunk 4]
```diff
@@ -27,9 +27,9 @@
 
 #if defined(HAVE_MAGMA) && defined(HAVE_OPENMP_OFFLOAD)
 #define MGMOL_PARALLEL_FOR(...)                                                \
-  _Pragma(MGMOL_STRINGIFY(omp target teams distribute parallel for is_device_ptr(__VA_ARGS__)))
+  _Pragma(MGMOL_STRINGIFY(omp target teams distribute parallel for schedule(static,1) is_device_ptr(__VA_ARGS__)))
```
Comment (jeanlucf22) [inline src/tools/memory_space.h]: this change doe not belong in this PR

File: tests/testTradeGhostValues.cc [hunk 5]
```diff
@@ -13,132 +13,186 @@
 
 #include "catch.hpp"
 
+// function of periodicity nx, ny, nz
```
Comment (jeanlucf22) [inline tests/testTradeGhostValues.cc]: this should not appear as a difference. please rebase

Comment (wjge) [reply tests/testTradeGhostValues.cc]: Something is wrong even after the rebase, maybe because I copied it from your fork before it is merged.
